File: app/src/main/java/com/example/newyorktimearticles/presentation/ui/ArticleListActivity.java
package com.example.newyorktimearticles.presentation.ui;

import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.View;
import android.widget.ProgressBar;

import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.recyclerview.widget.RecyclerView;

import com.example.newyorktimearticles.R;
import com.example.newyorktimearticles.data.ServerException;
import com.example.newyorktimearticles.data.model.Articles;
import com.example.newyorktimearticles.presentation.presenter.ArticlePresenter;
import com.example.newyorktimearticles.presentation.presenter.IArticlePresenter;
import com.google.android.material.snackbar.Snackbar;

import java.util.ArrayList;
import java.util.List;

public class ArticleListActivity extends AppCompatActivity implements OnArticleClickListener,
        IArticleView {

    public static final String ARTICLE_KEY = "selected_article";
    IArticlePresenter presenter;
    private ArticlesAdapter articleAdapter;
    private List<Articles> articlesList = new ArrayList<>();
    private ProgressBar progress;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_article_list);

        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        DrawerLayout mDrawerLayout = (DrawerLayout) findViewById(R.id.drawer);
        ActionBarDrawerToggle mDrawerToggle = new ActionBarDrawerToggle(this, mDrawerLayout, toolbar, 0, 0);
        mDrawerLayout.addDrawerListener(mDrawerToggle);
        mDrawerToggle.syncState();

        setUpRecyclerView();
        initProgress();

        presenter = new ArticlePresenter(this);
        presenter.fetchArticles();
    }

    private void initProgress() {
        progress = findViewById(R.id.progress);
    }

    private void setUpRecyclerView() {
        articleAdapter = new ArticlesAdapter(this, articlesList);
        RecyclerView recArticles = findViewById(R.id.recArticleList);
        recArticles.setAdapter(articleAdapter);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.article_list_menu, menu);
        return true;
    }

    @Override
    public void onArticleClick(Articles article) {
        Intent destination = new Intent(this, ArticleDetailActivity.class);
        destination.putExtra(ArticleListActivity.ARTICLE_KEY, article);
        startActivity(destination);
    }

    @Override
    public void showProgress(Boolean shouldShow) {
        if (shouldShow) progress.setVisibility(View.VISIBLE);
        else progress.setVisibility(View.GONE);
    }

    @Override
    public void getArticles(List<Articles> getArticlesList) {
        articleAdapter.updateArticleList(getArticlesList);
    }

    @Override
    public void displayError(ServerException e) {
        String errorMessage;
        switch (e) {
            case UnKnownHost:
                errorMessage = getString(R.string.no_internet_connection);
                break;
            case Other:
            default:
                errorMessage = getString(R.string.unknown_error);
                break;
        }
        Snackbar.make(progress.getRootView(), errorMessage, Snackbar.LENGTH_SHORT).show();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        presenter.onDestroy();
    }
}

File: app/src/main/java/com/gemidroid/newyorktimearticles/presentation/ui/ArticleListActivity.kt
package com.gemidroid.newyorktimearticles.presentation.ui

import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.View
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.drawerlayout.widget.DrawerLayout
import com.gemidroid.newyorktimearticles.R
import com.gemidroid.newyorktimearticles.presentation.viewmodel.ArticleListViewModel
import com.gemidroid.newyorktimearticles.utility.Constants.ARTICLE_KEY
import com.google.android.material.snackbar.Snackbar
import kotlinx.android.synthetic.main.activity_article_list.*
import org.koin.androidx.viewmodel.ext.android.viewModel
import java.net.UnknownHostException
import java.util.*

class ArticleListActivity : AppCompatActivity() {

    private val articleListViewModel by viewModel<ArticleListViewModel>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_article_list)

        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)

        val mDrawerLayout = findViewById<DrawerLayout>(R.id.drawer)
        val mDrawerToggle = ActionBarDrawerToggle(this, mDrawerLayout, toolbar, 0, 0)
        mDrawerLayout.addDrawerListener(mDrawerToggle)
        mDrawerToggle.syncState()

        getArticlesList()
        displayError()
    }


    private fun getArticlesList() {
        articleListViewModel.observeOnArticleList.observe(this, {
            progress.visibility = View.GONE
            rec_articles.apply {
                adapter = ArticlesAdapter(it) {
                    val destination =
                        Intent(this@ArticleListActivity, ArticleDetailActivity::class.java)
                    destination.putExtra(ARTICLE_KEY, it)
                    startActivity(destination)
                }
            }
        })
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.article_list_menu, menu)
        return true
    }

    private fun displayError() {
        articleListViewModel.observeOnError.observe(this, {
            progress.visibility = View.GONE
            val errorMessage: String = when (it) {
                is UnknownHostException -> getString(R.string.no_internet_connection)
                else -> getString(R.string.unknown_error)
            }
            Snackbar.make(progress!!.rootView, errorMessage, Snackbar.LENGTH_SHORT).show()

        })
    }
}